File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // 최대 이익 0, 저점은 임시로 첫번째 값으로 지정
        int maxProfit = 0, minPrice = prices[0];
        
        // 현재 값을 기준으로 우측으로 차례대로 이동하면서 계산
        for(int price : prices) {
            // 저점 갱신
            minPrice = Math.min(minPrice, price);
            
            // 최대 이익 갱신
            maxProfit = Math.max(maxProfit, price - minPrice);
        }
        
        return maxProfit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        var maxProfit = 0
        
        // 임시로 첫번째 값을 최소값으로 지정
        var minPrice = prices[0]
        
        // 현재 값을 기준으로 우측으로 차례대로 이동하면서 계산
        for(price in prices) {
            // 지금까지의 저점 계산
            minPrice = minPrice.coerceAtMost(price)
            
            // 최대 이익 갱신
            maxProfit = maxProfit.coerceAtLeast(price - minPrice)
        }
        
        return maxProfit
    }
}